plugins/python/python_plugin.c

```diff
@@ -1297,29 +1288,49 @@ void uwsgi_python_init_apps() {
 
 }
 
-void uwsgi_python_master_fixup(int step) {
 
-	static int master_fixed = 0;
-	static int worker_fixed = 0;
+void uwsgi_python_pre_uwsgi_fork() {
+	if (!uwsgi.master_process) return;
+
+	if (uwsgi.has_threads) {
+		// Since we are not calling cPython's fork, we need to avoid forking
+		// while the gil and import lock are held by a thread other than the
+		// main thread.
+		UWSGI_GET_GIL
+		_PyImport_AcquireLock();
+	}
+}
+
 
+void uwsgi_python_post_uwsgi_fork(int step) {
 	if (!uwsgi.master_process) return;
 
 	if (uwsgi.has_threads) {
 		if (step == 0) {
-			if (!master_fixed) {
-				UWSGI_RELEASE_GIL;
-				master_fixed = 1;
-			}
-		}	
+			// release locks witin master process
```

Comment:
```suggestion
			// release locks within master process
```

Reply:
can we also add a "why" here?

just enough to say we are releasing the locks acquired by `pre_uwsgi_fork` before forking so that Python can ....